### PR TITLE
Module 4

### PR DESCRIPTION
**Pull request by em4go**

### CODE REVIEW

File: src/module_4/model_selection.ipynb
Comment (sergiorozada12) [inline src/module_4/model_selection.ipynb:1]: Things about the NB. Most of them will look as nits to you. Maybe they are, but having the small things right is what makes the difference. People will notice them for sure, for the good and for the bad:

* If you set ```;``` after ```df.info()``` you avoid printing trash. Look at the output of ```baseline_model.fit(X_train[baseline_features], y_train)```.
* Check out namings. The function ```get_relevant_dataframe``` is a filter, set a proper name.
* Objects like ```order_sizes``` are dataframes, add ```df_``` to them.
* Check typos -> "and features and definde de target variable."
* This is a join, why so complicated?:

```
big_enough_orders = order_sizes[order_sizes >= min_products].index
return df[df.order_id.isin(big_enough_orders)].copy()
```

* If this is a filter ```.pipe(get_relevant_dataframe)```, why not using the native ```.filter``` method?
* You are not creating new columns here ```.assign(created_at=lambda x: pd.to_datetime(x.created_at))```, why aren't ```.apply```?
* Take your time to label the plots adequately.
* Sklearn has tools to deal with grid searches, and other searches, that can be more suitable than:

``` 
n_estimators = [20, 50, 100]
for learning_rate in [0.1, 0.05]:
    for depth in [1, 3, 5]:
        for n in n_estimators:
```

* I like the baselines, but did you take all the decisions with or without realizing the problem was imbalanced? Can you think of other baselines for imbalanced problems?
 * As a follow up, have you think how would you tackle the imbalancement problem?

File: src/module_4/model_fit.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import json
+from push_model import PushModel
+from utils import load_data, create_model_path, load_model
+import pandas as pd
+
+
+def handler_fit(event, _):
```
Comment (sergiorozada12) [inline src/module_4/model_fit.py:7]: Typings

File: src/module_4/model_fit.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    df = load_data()
+    df_train, _ = model.feature_label_split(df)
+
+    model.fit(df_train)
+
+    try:
+        model_path = create_model_path("push_model")
+        model.save(model_path)
+    except Exception as e:
+        return {
```
Comment (sergiorozada12) [inline src/module_4/model_fit.py:22]: Check out dataclasses, far more elegant to handle this responses

File: src/module_4/model_fit.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        "statusCode": "200",
+        "body": json.dumps(
+            {
+                "model_path": [model_path],
+            }
+        ),
+    }
+
+
+def handler_predict(event, _):
```
Comment (sergiorozada12) [inline src/module_4/model_fit.py:41]: Typings

File: src/module_4/push_model.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
+from sklearn.ensemble import GradientBoostingClassifier
+import joblib
+import pandas as pd
+from typing import Dict, Tuple
+
+
+class PushModel:
```
Comment (sergiorozada12) [inline src/module_4/push_model.py:7]: I love this class, very elegant way to save things

File: src/module_4/utils.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+def load_data(data_path: str = DATA_PATH) -> pd.DataFrame:
+    return pd.read_csv(data_path)
+
+
+def create_model_path(model_folder_path: str, model_name: str = "push.joblib") -> str:
+    model_name_without_extension = Path(model_name).stem
+    extension = ".joblib"
+
+    model_path = Path(model_folder_path) / model_name_without_extension + extension
```
Comment (sergiorozada12) [inline src/module_4/utils.py:18]: I think f-strings are more elegant to do this

File: src/module_4/utils.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+    if model_path.exists():
+        date = datetime.now().strftime("%yyyy_%mm_%dd")
+        model_path = (
+            Path(model_folder_path) / f"{model_name_without_extension}_{date}"
+            + extension
```
Comment (sergiorozada12) [inline src/module_4/utils.py:24]: Same here, why not at least

f"{model_name_without_extension}_{date}.joblib"

or 

f"{model_name_without_extension}_{date}{extension}"

**Review by sergiorozada12 — APPROVED** [review]
Good PR, and good job. You can merge, but take a look at the comments. It is very important to get the small things right.
